narwhal/types/src/tests/primary_type_tests.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use config::WorkerId;
+use crypto::{KeyPair, Signature};
+use fastcrypto::{
+    hash::{Digest, Hash},
+    traits::KeyPair as _,
+};
+use once_cell::sync::OnceCell;
+use proptest::{collection, prelude::*, strategy::Strategy};
+use rand::{rngs::StdRng, SeedableRng};
+use signature::Signer;
+
+use crate::{BatchDigest, CertificateDigest, Header, HeaderDigest, Metadata};
+
+fn arb_keypair() -> impl Strategy<Value = KeyPair> {
+    (any::<[u8; 32]>())
+        .prop_map(|rand| {
+            let mut rng = StdRng::from_seed(rand);
+            KeyPair::generate(&mut rng)
+        })
+        .no_shrink()
+}
+
+fn clean_signed_header(kp: KeyPair) -> impl Strategy<Value = Header> {
+    (
+        any::<u64>(),
+        any::<u64>(),
+        collection::vec((BatchDigest::arbitrary(), any::<u32>()), 0..10),
+        collection::vec(CertificateDigest::arbitrary(), 0..10),
+    )
+        .prop_map(move |(round, epoch, batches, parents)| {
+            let payload = batches
+                .into_iter()
+                .map(|(batch_digest, worker_id)| (batch_digest, worker_id as WorkerId))
+                .collect();
+
+            let parents = parents.into_iter().collect();
+
+            let header = Header {
+                author: kp.public().clone(),
+                round,
+                epoch,
+                payload,
+                parents,
+                id: OnceCell::default(),
+                signature: Signature::default(),
+                metadata: Metadata::default(),
+            };
+            Header {
+                id: OnceCell::with_value(header.digest()),
+                signature: kp.sign(Digest::from(header.digest()).as_ref()),
+                ..header
+            }
+        })
+}
+
+fn arb_signed_header(kp: KeyPair) -> impl Strategy<Value = Header> {
+    (
+        clean_signed_header(kp.copy()),
+        HeaderDigest::arbitrary(),
+        any::<usize>(),
+    )
+        .prop_map(move |(clean_header, random_digest, naughtiness)| {
+            let naughtiness = naughtiness % 100;
+            if naughtiness < 95 {
+                clean_header
+            } else if naughtiness < 99 {
+                let signature = kp.sign(Digest::from(random_digest).as_ref());
+                // naughty: we provide a well-signed random header
+                Header {
+                    id: OnceCell::with_value(random_digest),
+                    signature,
```

Comment:
How is this changing our test cases? The `assert` we do in the end is based on the `id() == digest()` right? So don't see anywhere the participation of the `signature` here

Reply:
The arbitrary generator is intended to be a general `Strategy` to produce a `Header` in tests, not necessarily limited to this test case.